drivers/pcie/host/pcie.c

```diff
@@ -49,55 +49,67 @@ void pcie_set_cmd(pcie_bdf_t bdf, uint32_t bits, bool on)
 
 bool pcie_get_mbar(pcie_bdf_t bdf, unsigned int index, struct pcie_mbar *mbar)
 {
-	bool valid_bar = false;
+	uintptr_t phys_addr;
 	uint32_t reg;
-	uintptr_t phys_addr = PCIE_CONF_BAR_NONE;
-	size_t size = 0;
+	size_t size;
 
-	for (reg = PCIE_CONF_BAR0; reg <= PCIE_CONF_BAR5; reg++) {
-		uint32_t addr = pcie_conf_read(bdf, reg);
-
-		/* skip useless bars and I/O Bars */
-		if (addr == 0xFFFFFFFFU || addr == 0x0U || PCIE_CONF_BAR_IO(addr)) {
-			continue;
-		}
-
-		if (index == 0) {
-			valid_bar = true;
-			break;
-		}
+	for (reg = PCIE_CONF_BAR0;
+	     index > 0 && reg <= PCIE_CONF_BAR5; reg++, index--) {
+		uintptr_t addr = pcie_conf_read(bdf, reg);
 
 		if (PCIE_CONF_BAR_64(addr)) {
 			reg++;
 		}
```

Comment:
```suggestion
		if (PCIE_CONF_BAR_MEM(addr) && PCIE_CONF_BAR_64(addr)) {
			reg++;
		}
```
when it is an I/O Bar bit 3 is already part of the address -> it might be 1 even though I/O Bars can not be 64 bit